ykrt/src/compile/jitc_yk/trace_builder.rs

```diff
@@ -615,12 +642,115 @@ impl<'a> TraceBuilder<'a> {
+        let jit_type_idx = self.handle_type(test_val.type_(self.aot_mod))?;
+        let jit_int_type = match self.jit_mod.type_(jit_type_idx) {
+            jit_ir::Ty::Integer(it) => it,
+            _ => unreachable!(),
+        }
+        .to_owned();
+
+        // Find out which case we traced.
+        let guard = match case_dests.iter().position(|&cd| cd == next_bb.bb_idx()) {
+            Some(cidx) => {
+                // A non-default case was traced.
+                let val = case_values[cidx];
+                let bb = case_dests[cidx];
+
+                // Build the constant value to guard.
+                let jit_const = jit_int_type
+                    .to_owned()
+                    .make_constant(&mut self.jit_mod, val)?;
+                let jit_const_opnd = jit_ir::Operand::Const(self.jit_mod.const_idx(&jit_const)?);
+
+                // Perform the comparison.
+                let jit_test_val = self.handle_operand(test_val)?;
+                let cmp_inst =
+                    jit_ir::IcmpInst::new(jit_test_val, jit_ir::Predicate::Equal, jit_const_opnd);
+                let jit_cond = self.jit_mod.push_and_make_operand(cmp_inst.into())?;
+
+                // Guard the result of the comparison.
+                self.create_guard(&jit_cond, bb == next_bb.bb_idx(), sm)?
+            }
+            None => {
+                // The default case was traced.
+                //
+                // We need a guard that expresses that `test_val` was not any of the `case_values`.
+                // We encode this in the JIT IR like:
+                //
+                //     member = case_val1 || ... || case_valN
+                //     guard member == 0
+                //
+                // OPT: This could be much more efficient if we introduced a special "none of these
+                // values" guard. It could codegen the comparisons into a bitfield and efficiently
+                // bitwise OR a whole word's worth of comparison outcomes at once.
+                //
+                // OPT: Also depending on the shape of the cases you may be able to optimise. e.g.
+                // if they are a consecutive run, you could do a range check instead of all of
+                // these comparisons.
+                let mut cmps_opnds = Vec::new();
+                for cv in case_values {
+                    // Build a constant of the case value.
+                    let jit_const = jit_int_type
+                        .to_owned()
+                        .make_constant(&mut self.jit_mod, *cv)?;
+                    let jit_const_opnd =
+                        jit_ir::Operand::Const(self.jit_mod.const_idx(&jit_const)?);
+
+                    // Do the comparison.
+                    let jit_test_val = self.handle_operand(test_val)?;
+                    let cmp = jit_ir::IcmpInst::new(
+                        jit_test_val,
+                        jit_ir::Predicate::Equal,
```

Comment:
I was expecting "not equal" here because that gives a slightly more intuitive notion of why the guard exists. It's not a big deal though.

Reply:
We compare all of the cases that weren't taken, or their 0/1 decisions together and guard that *that* was false.

You could apply demorgan's law to invert the logic, but me and lukas agreed that this way around is less confusing.